core/src/msg.rs

```diff
@@ -0,0 +1,352 @@
+        }
+
+        // Return the found message (without removing)
+        if let Some(mtype) = candidate_type {
+            self.messages
+                .get(&mtype)
+                .and_then(|msgs| msgs.first())
+                .map(|msg| (mtype, &msg.data[..]))
+        } else {
+            None
+        }
+    }
+
+    /// Get total number of messages in the queue (for MSG_COPY)
+    pub fn get_total_message_count(&self) -> usize {
+        self.messages.values().map(|msgs| msgs.len()).sum()
+    }
+
+    /// Get message by index (for MSG_COPY)
+    pub fn get_message_by_index(&self, index: usize) -> Option<&Message> {
+        let mut current_index = 0;
+
+        // Iterate over all messages in order of message type
+        for messages in self.messages.values() {
+            if index < current_index + messages.len() {
+                return messages.get(index - current_index);
+            }
+            current_index += messages.len();
+        }
+        None
+    }
+
+    /// Remove the first message of the specified type
+    pub fn remove_first_message_of_type(&mut self, mtype: i64) -> Option<Message> {
+        if let Some(messages) = self.messages.get_mut(&mtype) {
+            let removed_msg = messages.remove(0);
+
+            // Update statistics
+            self.total_bytes -= removed_msg.data.len();
+            self.msqid_ds.msg_cbytes -= removed_msg.data.len() as __kernel_size_t;
+            self.msqid_ds.msg_qnum -= 1;
+
+            // If the message list of this type is empty, remove the entire type entry
+            if messages.is_empty() {
+                self.messages.remove(&mtype);
+            }
+            return Some(removed_msg);
+        }
+        None
+    }
+
+    /// Remove the first message of the specified type with matching content
+    pub fn remove_matched_message_by_type(&mut self, mtype: i64, data: &[u8]) -> AxResult<()> {
+        if let Some(messages) = self.messages.get_mut(&mtype) {
+            // Find the first message with matching content and type
+            if let Some(pos) = messages.iter().position(|msg| msg.data == data) {
+                let removed_msg = messages.remove(pos);
+
+                // Update core queue statistics in the removal method
+                self.total_bytes -= removed_msg.data.len();
+                self.msqid_ds.msg_cbytes -= removed_msg.data.len() as __kernel_size_t;
+                self.msqid_ds.msg_qnum -= 1;
+
+                // If the message list of this type is empty, remove the entire type entry
+                if messages.is_empty() {
+                    self.messages.remove(&mtype);
+                }
+
+                return Ok(());
+            }
+        }
+        Err(AxError::NoMemory)
+    }
+}
+
+/// Message queue manager
+pub struct MsgManager {
+    /// key -> msqid mapping
+    key_msqid: BTreeMap<i32, i32>,
+    /// msqid -> message queue inner structure
+    pub msqid_inner: BTreeMap<i32, Arc<Mutex<MsgInner>>>,
```

Comment:
Consider remove this `pub`. I see that it's used in `MSG_STAT`. Add some iterator methods, for example `iter_msg_queues`. Then in `MSG_STAT`, write some Rust-style functional programming instead of C-style for-loop. Hint: `xxx.iter_msg_queues().filter(...).nth(...)`